File: 7_kyu/Remove_duplicate_words.py
def remove_duplicate_words(s: str) -> str:
    seen = set()
    def was_seen(x:str) -> bool:
        nonlocal seen
        if not (result := x in seen):
            seen.add(x)
        return result
    return " ".join( it for it in s.split() if not was_seen(it) )

File: 7_kyu/Remove_duplicate_words.go
package kata

import "strings"

func RemoveDuplicateWords(str string) string {
	input := strings.Fields(str)
	seen := make(map[string]bool, len(input))
	output := make([]string, 0, len(input))
	for _, s := range input {
		if !seen[s] {
			output = append(output, s)
			seen[s] = true
		}
	}
	return strings.Join(output, " ")
}
